src/swap/do_action.rs

```diff
@@ -0,0 +1,212 @@
+use crate::{swap::db, SwapId};
+use comit::Timestamp;
+
+#[async_trait::async_trait]
+pub trait Do<E>
+where
+    Self: CheckMemory<E> + ShouldAbort + Execute<E> + Remember<E>,
+    E: Clone + Send + Sync + 'static,
+    <Self as Execute<E>>::Args: Send + Sync,
```

Comment:
Putting trait bounds on traits in definition sites can be a footgun because you have to restate them at every declaration.

To avoid code duplication between several impls (if that is what you are going for), you could:

1. Define the trait without bounds
2. Make a private function that implements the actual logic
3. Implement the trait through blanket impls by forwarding to the private function

Reply:
The idea here is to only ever implement this trait once:
```rust
#[async_trait::async_trait]
impl<E, A> Do<E> for A
where
    A: CheckMemory<E> + ShouldAbort + Execute<E> + Remember<E>,
    E: Clone + Send + Sync + 'static,
    <Self as Execute<E>>::Args: Send + Sync,
{
}
```
This trait just composes those traits together in a consistent way.

Reply:
@thomaseizinger: after our conversation this morning I was convinced that replacing the trait with a free function was superior, but it does come with the downside that I have to add a much noisier trait bound per action to the `hbit_herc20` function (and `herc20_hbit` in the near future).

Since I don't really see harm in the trait itself and it turns out that I can use it to hide some details I am leaning towards keeping it. Maybe you have a better idea though :)

Reply:
Obviously completely up to you! :)

If that is the main justification for the trait now, then I'd still question if it is such a good design if you need to hide a lot of noise behind a trait :D

Exposing yourself to this noise/pain might yield a different and potentially better design that actually reduces the noise instead of just hiding it :)